pkgs/desktops/lxqt/obconf-qt/default.nix

```diff
@@ -11,6 +11,7 @@
 , lxqt-build-tools
 , openbox
 , lxqtUpdateScript
+, perl
 }:
 
 mkDerivation rec {
```

Comment:
How did the dependencies change when the version stayed the same?

Reply:
It may be related to other dependences which have been updated. `lxqt-build-tools` for instance has been updated. Without  `perl` in the dependence list, building fails at the configuration phase:

```
CMake Error at /nix/store/8kpnrqj3palxjvawyjrnfdmnig0b33wy-cmake-3.19.7/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find Perl (missing: PERL_EXECUTABLE)
Call Stack (most recent call first):
  /nix/store/8kpnrqj3palxjvawyjrnfdmnig0b33wy-cmake-3.19.7/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
  /nix/store/8kpnrqj3palxjvawyjrnfdmnig0b33wy-cmake-3.19.7/share/cmake-3.19/Modules/FindPerl.cmake:82 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  /nix/store/d4hlixz1vpahcy924a8bykqvp3qi1zj4-lxqt-build-tools-0.9.0/share/cmake/lxqt-build-tools/modules/LXQtTranslateDesktop.cmake:27 (find_package)
  CMakeLists.txt:46 (include)
```

It seems to be a requirement of the new version of `lxqt-build-tools`, related to translations.

I have tried using `propagatedNativeBuildInputs` hoping it existed and worked similar to `propagatedBuildInputs`, but with no success.

Reply:
In this case is it correct to have `perl` as a propagated build input in `lxqt-build-tools`? With that it is not needed to add `perl` as an explicit dependence in `obconf-qt`, which successfully builds.

Reply:
I have added the `perl` dependence to `propagatedBuildInputs` in `lxqt-build-tools`.

Reply:
Yeah, thats cleaner than adding it for every package. Can we maybe define the perl location or patch the shebang instead?

Reply:
@SuperSandro2000 do you mean something like the following (which also works)?

``` patch
diff --git a/pkgs/desktops/lxqt/lxqt-build-tools/default.nix b/pkgs/desktops/lxqt/lxqt-build-tools/default.nix
index 27fda636103..b7dff56774c 100644
--- a/pkgs/desktops/lxqt/lxqt-build-tools/default.nix
+++ b/pkgs/desktops/lxqt/lxqt-build-tools/default.nix
@@ -31,14 +31,22 @@ mkDerivation rec {
     qtbase
     glib
     pcre
-  ];
-
-  propagatedBuildInputs = [
     perl # needed by LXQtTranslateDesktop.cmake
   ];
 
   setupHook = ./setup-hook.sh;
 
+  postPatch = ''
+    # Explicitly define the perl binary location. Otherwise
+    # every package which uses lxqt-build-tools would have to find
+    # perl using cmake again, and for that it would need to depend on
+    # perl.
+
+    substituteInPlace cmake/modules/LXQtTranslateDesktop.cmake \
+      --replace "find_package(Perl REQUIRED)" "# find_package(Perl REQUIRED)" \
+      --replace "\''${PERL_EXECUTABLE}" "${perl}/bin/perl"
+  '';
+
   # We're dependent on this macro doing add_definitions in most places
   # But we have the setup-hook to set the values.
   postInstall = ''
-- 
2.31.1
```

I am not sure it is better than propagating the perl dependence, because it makes the nix expression a little more complicated. What do you think?

Reply:
Could you perhaps pass the perl executable path via cmakeFlagsArray? I see setup-hooks.sh already adds some stuff to this. Maybe it could be possible to add `-DPERL_LOCATION=...`. I’m not really sure, just trying to think of ways to eat the cake and have it too.

Reply:
Doing ``-DPERL_LOCATION=...`` would be ideal.

Reply:
Defining `-DPERL_LOCATION=...` seems to be helpless.

With the following
``` bash
    cmakeFlagsArray+=(
      -DPERL_EXECUTABLE=${perl}/bin/perl
    )
```
`perl` can be found. But I do not know how to define this only once (for `lxqt-build-tools`) and have it working for every packaging depending on `lxqt-build-tools` without explicitly depending on `perl`.

If I try to add the cmake flag to the array in the setup hook (a bash script), I do not have access to `${perl}`.  If anybody knows how to do it, please tell me.

Maybe I am misunderstanding the suggestions given in this discussion.


> Doing `-DPERL_LOCATION=...` would be ideal.

@SuperSandro2000 why would this be ideal? What is the problem with propagating the dependence on `perl`? In any case `perl` will have to be available for the packages that depend on `lxqt-build-tools`.

Reply:
> If I try to add the cmake flag to the array in the setup hook (a bash script), I do not have access to ${perl}. If anybody knows how to do it, please tell me.

Maybe @perl@ and substituteAll but that would be almost equivalent to patching the cmake script.

> why would this be ideal?

Patching should be avoided if we can use some flag or option. If we can't than this is alright, too.

> What is the problem with propagating the dependence on perl?

I can't find the details right now but it should be avoided if there are easy ways around it.

Reply:
For me I would keep the propagated dependence solution. It is the simplest and most readable, and it works.

Reply:
Sorry, I did not mean to delay things with dead-ends. The propagated build dependency doesn’t seem so bad to me personally.